File: lib/slatedb/write_batch.rb
# frozen_string_literal: true

module SlateDb
  class WriteBatch
    # Add a put operation to the batch.
    #
    # @param key [String] The key to store
    # @param value [String] The value to store
    # @param ttl [Integer, nil] Time-to-live in milliseconds
    # @return [self] Returns self for method chaining
    #
    # @example
    #   batch.put("key", "value")
    #   batch.put("key2", "value2", ttl: 60_000)
    #
    def put(key, value, ttl: nil)
      if ttl
        _put_with_options(key, value, { ttl: ttl })
      else
        _put(key, value)
      end
      self
    end

    # Add a delete operation to the batch.
    #
    # @param key [String] The key to delete
    # @return [self] Returns self for method chaining
    #
    # @example
    #   batch.delete("key")
    #
    def delete(key)
      _delete(key)
      self
    end

    # Add a merge operation to the batch.
    #
    # @param key [String] The key to merge into
    # @param value [String] The merge operand to apply
    # @param ttl [Integer, nil] Time-to-live in milliseconds
    # @return [self] Returns self for method chaining
    #
    # @example
    #   batch.merge("key", "part1")
    #   batch.merge("key", "part2", ttl: 30_000)
    #
    def merge(key, value, ttl: nil)
      if ttl
        _merge_with_options(key, value, { ttl: ttl })
      else
        _merge(key, value)
      end
      self
    end
  end
end

File: ext/slatedb/src/write_batch.rs
use std::cell::RefCell;

use magnus::prelude::*;
use magnus::{function, method, Error, RHash, Ruby};
use slatedb::config::{PutOptions, Ttl};
use slatedb::WriteBatch as SlateWriteBatch;

use crate::errors::invalid_argument_error;
use crate::utils::get_optional;

/// Ruby wrapper for SlateDB WriteBatch.
///
/// This struct is exposed to Ruby as `SlateDb::WriteBatch`.
#[magnus::wrap(class = "SlateDb::WriteBatch", free_immediately, size)]
pub struct WriteBatch {
    inner: RefCell<SlateWriteBatch>,
}

impl WriteBatch {
    /// Create a new empty WriteBatch.
    pub fn new() -> Self {
        Self {
            inner: RefCell::new(SlateWriteBatch::new()),
        }
    }

    /// Add a put operation to the batch.
    pub fn put(&self, key: String, value: String) -> Result<(), Error> {
        if key.is_empty() {
            return Err(invalid_argument_error("key cannot be empty"));
        }

        self.inner
            .borrow_mut()
            .put(key.as_bytes(), value.as_bytes());

        Ok(())
    }

    /// Add a put operation with options to the batch.
    ///
    /// Options:
    /// - ttl: Time-to-live in milliseconds
    pub fn put_with_options(&self, key: String, value: String, kwargs: RHash) -> Result<(), Error> {
        if key.is_empty() {
            return Err(invalid_argument_error("key cannot be empty"));
        }

        let ttl = get_optional::<u64>(&kwargs, "ttl")?;
        let put_opts = PutOptions {
            ttl: match ttl {
                Some(ms) => Ttl::ExpireAfter(ms),
                None => Ttl::Default,
            },
        };

        self.inner
            .borrow_mut()
            .put_with_options(key.as_bytes(), value.as_bytes(), &put_opts);

        Ok(())
    }

    /// Add a delete operation to the batch.
    pub fn delete(&self, key: String) -> Result<(), Error> {
        if key.is_empty() {
            return Err(invalid_argument_error("key cannot be empty"));
        }

        self.inner.borrow_mut().delete(key.as_bytes());

        Ok(())
    }

    /// Take ownership of the inner WriteBatch (consumes it).
    /// Used internally when writing the batch to the database.
    pub fn take(&self) -> Result<SlateWriteBatch, Error> {
        Ok(self.inner.replace(SlateWriteBatch::new()))
    }
}

/// Define the WriteBatch class on the SlateDb module.
pub fn define_write_batch_class(ruby: &Ruby, module: &magnus::RModule) -> Result<(), Error> {
    let class = module.define_class("WriteBatch", ruby.class_object())?;

    // Class methods
    class.define_singleton_method("new", function!(WriteBatch::new, 0))?;

    // Instance methods
    class.define_method("_put", method!(WriteBatch::put, 2))?;
    class.define_method(
        "_put_with_options",
        method!(WriteBatch::put_with_options, 3),
    )?;
    class.define_method("_delete", method!(WriteBatch::delete, 1))?;

    Ok(())
}
